interbotix_common_toolbox/interbotix_gravity_compensation/src/interbotix_gravity_compensation_obj.cpp

```diff
@@ -0,0 +1,475 @@
+  joint_group_pub_ = this->create_publisher<interbotix_xs_msgs::msg::JointGroupCommand>(
+    "commands/joint_group", 10);
+
+  // Create the Service for enabling/disabling gravity compensation
+  gravity_compensation_enable_srv_ = this->create_service<std_srvs::srv::SetBool>(
+    "gravity_compensation_enable",
+    std::bind(
+      &InterbotixGravityCompensation::gravity_compensation_enable_cb, this,
+      std::placeholders::_1, std::placeholders::_2));
+
+  // Create the client for the 'OperatingModes' service
+  operating_modes_client_ = this->create_client<interbotix_xs_msgs::srv::OperatingModes>(
+    "set_operating_modes");
+
+  // Create the client for the 'TorqueEnable' service
+  torque_enable_client_ = this->create_client<interbotix_xs_msgs::srv::TorqueEnable>(
+    "torque_enable");
+
+  // Wait for the 'OperatingModes' service to be available
+  while (!operating_modes_client_->wait_for_service(std::chrono::seconds(1))) {
+    if (!rclcpp::ok()) {
+      RCLCPP_ERROR(
+        this->get_logger(),
+        "Interrupted while waiting for %s. Exiting.",
+        operating_modes_client_->get_service_name());
+      success = false;
+      return;
+    }
+    RCLCPP_INFO(
+      this->get_logger(),
+      "%s not available, waiting again...",
+      operating_modes_client_->get_service_name());
+  }
+
+  // Wait for the 'TorqueEnable' service to be available
+  while (!torque_enable_client_->wait_for_service(std::chrono::seconds(1))) {
+    if (!rclcpp::ok()) {
+      RCLCPP_ERROR(
+        this->get_logger(),
+        "Interrupted while waiting for %s. Exiting.",
+        torque_enable_client_->get_service_name());
+      success = false;
+      return;
+    }
+    RCLCPP_INFO(
+      this->get_logger(),
+      "%s not available, waiting again...",
+      torque_enable_client_->get_service_name());
+  }
+
+  // Load the motor specs
+  if (!load_motor_specs(motor_specs)) {
+    success = false;
+    return;
+  }
+
+  // Prepare for the inverse dynamics solver
+  if (!prepare_tree()) {
+    success = false;
+    return;
+  }
+
+  // Prompt the user to enable gravity compensation
+  RCLCPP_INFO(
+    this->get_logger(),
+    "Gravity compensation node is up."
+  );
+
+  // Warn the user that the joints will be turned off temporarily
+  RCLCPP_WARN(
+    this->get_logger(),
+    "WARNING: The joints will be turned off temporarily while enabling/disabling "
+    "gravity compensation. Please make sure the robot is in a safe position or "
+    "manually held before enabling gravity compensation."
+  );
+}
+
+void InterbotixGravityCompensation::joint_state_cb(
+  const sensor_msgs::msg::JointState::SharedPtr msg)
+{
+  // If gravity compensation is not enabled, return
+  if (!gravity_compensation_enabled_) {
+    return;
+  }
+
+  // Construct the id solver
+  KDL::TreeIdSolver_RNE idsolver(tree_, KDL::Vector(0, 0, -9.81));
+
+  // Set the joint positions and velocities
+  for (size_t i = 0; i < msg->position.size(); i++) {
```

Comment:
Should we only go up to num_joints_arm_ here? Assuming they're the first joints in the arm.

Reply:
Here the variables all have 9 joints as defined in the URDF, i.e., 6 arm joints, 1 gripper, and 2 fingers. The KDL tree is built based on the UDRF. And the joint state message also matches that.